### DISCUSSION TITLE
Multiple repos vs monorepo approaches

### DISCUSSION THREAD

**Discussion by stanislaw** [nasa/cFS#330] [Q&A]
Hello,

I would like to ask my question based on what Jacob wrote in the recent thread about SCH:

> According to the sch 2.2.2 documentation at https://github.com/nasa/SCH/blob/master/docs/SCH_v2.2.2_VersionDescriptionDocument.pdf it is compatible with cFE 6.6.  There is an open source release expected in the near future for sch to update for compatibility with 7.0 cFE, which will be part of the Caelum release bundle.  Until it’s released, you will run into issues trying to build sch 2.2.2 to run with the Caelum development branch bundle.  Typically I think the open source users either pick an older version of the bundle for which the apps have compatible releases, or they go through the effort of updating the app to whatever bundle version they are using.
 
> We realize the out of sync releases aren’t ideal and have put significant effort into expediting app updates as well as the release processes, but there are steps we are required to take for an open source release that are unfortunately not within our schedule control.  We will definitely continue to push this on our side.  We also try to keep the APIs as stable as possible for minor releases, but Caelum is a major release so there were impacts across all the apps based on major cleanup and refactoring activities this round.
 
> We are working very hard on a core set of GSFC apps that we plan to release open source in the near future with Caelum compatibility…

I was wondering if you considered a monolithic cFS repository approach where all core and most used apps would be included into one tree?

A clear advantage of a monorepo approach is that all included software shares the same release lifecycle and with some amount of integration tests, the whole tree could be kept consistent in terms of the functionality and the interfaces. The additional benefit, I guess, is that the release management for the cFS team could be simpler this way.

I would be curious to hear what was the idea behind the cFS development team's decision for splitting the core cFS repositories?

P.S. In my understanding, this topic is also connected to another one that I opened before: [In-source vs out-of-source use of cFS: Recommended project structure? #319](https://github.com/nasa/cFS/discussions/319).

Thanks.

**Comment by jbohren** [comment]
It's possible that export control / intellectual property review is the real bottleneck. If that's the case then releasing everything as a single unit will just be slowed down to the most difficult-to-review subunit.

I think creating a monorepo also has the potential to increase the upstream contribution burden. Users would have to clone that whole repo, and would want to keep application-specific code out of it if they ever want to upstream bug fixes or feature proposals.

**Reply by skliper** [reply]
You've hit on some of the prime reasons for the distributed repo model.  While it certainly poses challenges related to certain repos falling behind (either due to release challenges or lack of resources to update the low-use apps), we are able to maintain/update/release the other elements based on their readiness vs having everything updated.  It's especially useful to have the bundle/framework updates available so app developers and users can more easily plan for compatibility, estimate impacts, trade versions, collaborate/contribute, etc.  Note https://github.com/OpenSatKit/OpenSatKit does provide a functional "distribution" that includes compatible versions of many of the most popular apps along with a compatible framework.
